### PR TITLE
Update interface to version 3.

### PR DESCRIPTION
**Pull request by jakobnissen**
Ref #140 , #160 . Maybe needs to be merged with #160 .

This is a fairly large rewrite of BioSequences. The changes are breaking, and so needs to go into v3, but they appear bigger than they are. The large majority are internal changes.

## API changes
`Alphabet` and `BioSequence` are now defined as interfaces. They have a small, minimal interface they *must* implement. You can find it in the docstrings in this PR for the two types - it's nearly identical to those discussed in #140 and #160. This is breaking, but unlikely to actually break anyone's code because it was difficult to actually implement these new types beforehand.

`inbounds_getindex` has been removed. This is part of a larger plan to make indexing more like vectors for v3.

## Internal changes only
[ to be added ]

### CODE REVIEW

**Comment by jakobnissen** [comment]
Ah, of course. It fails because it requires the latest master branch version of BioSymbols. The latest master contains breaking changes.

**Review by CiaranOMara — COMMENTED** [review]
My comments pertain to the documentation and nomenclature, the code appears solid. However, there are some inconsistencies in nomenclature. The following terms are used interchangeably:
* coding units/elements
* symbol lists/sets
* element/symbol/data

It is not always clear what is happening. Sometimes the general terms "element" or "data" are used in an abstract description. The problem with this is that we lose information about whether the element or data is a symbol that is encoded, or an encoding that is packed.

It may be easier to tighten the nomenclature with a separate pull request?

File: src/biosequence/biosequence.jl [hunk 1]
```diff
@@ -4,78 +4,129 @@
 ### This file is a part of BioJulia.
 ### License is MIT: https://github.com/BioJulia/BioSequences.jl/blob/master/LICENSE.md
 
+"""
+    BioSequence{A <: Alphabet}
+
+`BioSequence` is the main abstract type of `BioSequences`.
+Its subtypes are characterized by:
+* Being a linear container type with random access and indices `Base.OneTo(length(x))`.
+* Containing zero or more coding elements of type `encoded_data_eltype(typeof(x))`.
```
[the remark this answers is not in the document]
Comment (jakobnissen) [reply src/biosequence/biosequence.jl:13]: Here, I think it's better to be explicit that the BioSequence contains the internal representation, so I've changed it to "* Containing zero or internal data elements of type `encoded_data_eltype(typeof(x))`."

File: test/alphabet.jl [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+* `decode(::A, ::X)::E` decodes an `X` to an element `E`.
+* Except for `eltype` which must follow Base conventions, all functions operating
+on `Alphabet` should operate on instances of the alphabet, not the type.
+
+
+If you want interoperation with existing subtypes of `BioSequence`,
+the element type `E` must be `UInt`, and you must also implement:
+* `BitsPerSymbol(::A)::BitsPerSymbol{N}`, where the `N` must be zero
+or a power of two in [1, 2, 4, 8, 16, 32, [64 for 64-bit systems]].
+"""
```
[the remark this answers is not in the document]
Comment (jakobnissen) [reply test/alphabet.jl:66]: Whooops, yeah, I used the docstring to make sure the tests actually tested for the docs, and then I just forgot to remove it afterwards. Good catch.

File: src/bit-manipulation/bitindex.jl [hunk 3]
```diff
@@ -59,7 +71,7 @@ end
 Base.show(io::IO, i::BitIndex) = print(io, '(', index(i), ", ", offset(i), ')')
 
 "Extract the element stored in a packed bitarray referred to by bidx."
-@inline function extract_encoded_element(bidx::BitIndex{N,W}, data::AbstractArray{W}) where {N,W}
+@inline function extract_encoded_element(bidx::BitIndex{N,W}, data::Union{AbstractArray{W}, NTuple{T, W}}) where {N,W,T}
```
[the remark this answers is not in the document]
Comment (jakobnissen) [reply src/bit-manipulation/bitindex.jl:74]: Hm, not sure. One of the salient things should be that it extracts the encoded version of a single element.

**Comment by jakobnissen** [comment]
Thanks for the review! You are right. Except for the two points I commented, I have made the requested changes.

**Comment by BenJWard** [comment]
I'm happy for this to get merged first and then if there's anything not made moot in #160 it can be merged afterwards.

**Comment by BenJWard** [comment]
@jakobnissen I've released BioSymbols v5, and made some commits to your interfaces branch to get tests to work.

One of the test failures was due to struct definitions in `@testset` blocks not being permitted in julia 1.0, so I tweaked that.

Tests work for julia 1 tests, but not for 1.0 now. The current reason seems to be a method ambiguity between the constructor on line 67 of longsequence/subseq.jl and the constructor on line 26 of longsequence/conversion.jl

I don't know what you plan to do about this, one solution I've found is to change the sebseq.jl function to:

```julia
function LongSequence{A}(seq::LongSubSeq{A}) where {A<:NucleicAcidAlphabet}
	_copy_seqview(LongSequence{A}, seq)
end
```

Which as far as I can tell achieves the same as the current 
```julia
function (::Type{T})(seq::LongSubSeq{<:NucleicAcidAlphabet{N}}) where
{N, T<:LongSequence{<:NucleicAcidAlphabet{N}}}
	_copy_seqview(T, seq)
end
```

But without the ambiguity.

**Comment by BenJWard** [comment]
All tests are passing now for 1 and 1.0

**Comment by BenJWard** [comment]
I'd like to merge this, this evening into v3 if that's alright.

**Pull request merged**